File: language_policy_test.py
#!/usr/bin/env python3

"""
Language Policy Backend Test for RV Service Desk

Tests the core language policy functionality to ensure:
1. EN mode never outputs translation blocks
2. RU/ES/AUTO modes output translation blocks when appropriate
3. Language policy is enforced declaratively, not through prompt hacks
4. All language policy functions work as expected
"""

import sys
import json
import subprocess
import os
from pathlib import Path

def run_node_test(test_code):
    """Run Node.js test code and return result"""
    # Write test code to temp file
    test_file = Path("/tmp/lang_test.mjs")
    test_file.write_text(test_code)
    
    # Run with Node.js
    result = subprocess.run(
        ["node", str(test_file)], 
        capture_output=True, 
        text=True, 
        cwd="/app"
    )
    
    # Clean up
    test_file.unlink(missing_ok=True)
    
    return {
        "success": result.returncode == 0,
        "stdout": result.stdout.strip(),
        "stderr": result.stderr.strip(),
        "returncode": result.returncode
    }

class LanguagePolicyTester:
    def __init__(self):
        self.tests_run = 0
        self.tests_passed = 0
        self.failures = []

    def test(self, name, test_func):
        """Run a single test"""
        self.tests_run += 1
        print(f"\n🔍 Testing {name}...")
        
        try:
            result = test_func()
            if result:
                self.tests_passed += 1
                print(f"✅ Passed")
                return True
            else:
                print(f"❌ Failed")
                self.failures.append(name)
                return False
        except Exception as e:
            print(f"❌ Failed with error: {str(e)}")
            self.failures.append(f"{name}: {str(e)}")
            return False

    def test_resolve_language_policy_en_mode(self):
        """Test: EN mode returns includeTranslation=false"""
        test_code = '''
        import { resolveLanguagePolicy } from "./src/lib/lang.js";
        
        const policy = resolveLanguagePolicy("EN", "EN");
        
        if (policy.mode !== "EN") {
            console.error("Expected mode EN, got:", policy.mode);
            process.exit(1);
        }
        
        if (policy.includeTranslation !== false) {
            console.error("Expected includeTranslation false, got:", policy.includeTranslation);
            process.exit(1);
        }
        
        if (policy.translationLanguage !== undefined) {
            console.error("Expected translationLanguage undefined, got:", policy.translationLanguage);
            process.exit(1);
        }
        
        console.log("EN mode correctly returns includeTranslation=false");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_resolve_language_policy_ru_mode(self):
        """Test: RU mode returns includeTranslation=true"""
        test_code = '''
        import { resolveLanguagePolicy } from "./src/lib/lang.js";
        
        const policy = resolveLanguagePolicy("RU", "RU");
        
        if (policy.mode !== "RU") {
            console.error("Expected mode RU, got:", policy.mode);
            process.exit(1);
        }
        
        if (policy.includeTranslation !== true) {
            console.error("Expected includeTranslation true, got:", policy.includeTranslation);
            process.exit(1);
        }
        
        if (policy.translationLanguage !== "RU") {
            console.error("Expected translationLanguage RU, got:", policy.translationLanguage);
            process.exit(1);
        }
        
        console.log("RU mode correctly returns includeTranslation=true with translationLanguage=RU");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_resolve_language_policy_auto_en(self):
        """Test: AUTO+EN returns includeTranslation=false"""
        test_code = '''
        import { resolveLanguagePolicy } from "./src/lib/lang.js";
        
        const policy = resolveLanguagePolicy("AUTO", "EN");
        
        if (policy.mode !== "AUTO") {
            console.error("Expected mode AUTO, got:", policy.mode);
            process.exit(1);
        }
        
        if (policy.includeTranslation !== false) {
            console.error("Expected includeTranslation false, got:", policy.includeTranslation);
            process.exit(1);
        }
        
        console.log("AUTO+EN correctly returns includeTranslation=false");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_resolve_language_policy_auto_ru(self):
        """Test: AUTO+RU returns includeTranslation=true"""
        test_code = '''
        import { resolveLanguagePolicy } from "./src/lib/lang.js";
        
        const policy = resolveLanguagePolicy("AUTO", "RU");
        
        if (policy.mode !== "AUTO") {
            console.error("Expected mode AUTO, got:", policy.mode);
            process.exit(1);
        }
        
        if (policy.includeTranslation !== true) {
            console.error("Expected includeTranslation true, got:", policy.includeTranslation);
            process.exit(1);
        }
        
        if (policy.translationLanguage !== "RU") {
            console.error("Expected translationLanguage RU, got:", policy.translationLanguage);
            process.exit(1);
        }
        
        console.log("AUTO+RU correctly returns includeTranslation=true with translationLanguage=RU");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_validate_final_report_en_mode_english_only(self):
        """Test: EN mode (includeTranslation=false) passes for English-only report"""
        test_code = '''
        import { validateFinalReportOutput } from "./src/lib/mode-validators.js";
        
        const englishOnlyReport = "Water pump not operating per spec. Labor: 1.0 hr. Total labor: 1.0 hr.";
        const result = validateFinalReportOutput(englishOnlyReport, false);
        
        if (!result.valid) {
            console.error("Expected valid=true, got:", result.valid);
            console.error("Violations:", result.violations);
            process.exit(1);
        }
        
        console.log("EN mode correctly validates English-only report");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_validate_final_report_en_mode_with_translation(self):
        """Test: EN mode rejects bilingual report with translation"""
        test_code = '''
        import { validateFinalReportOutput } from "./src/lib/mode-validators.js";
        
        const bilingualReport = "Water pump not operating per spec. Labor: 1.0 hr.\\n\\n--- TRANSLATION ---\\n\\nНасос не работает. Работа: 1.0 час.";
        const result = validateFinalReportOutput(bilingualReport, false);
        
        if (result.valid) {
            console.error("Expected valid=false for bilingual report in EN mode, got:", result.valid);
            process.exit(1);
        }
        
        const hasLanguagePolicyViolation = result.violations.some(v => v.includes("EN mode must not include"));
        if (!hasLanguagePolicyViolation) {
            console.error("Expected language policy violation, got violations:", result.violations);
            process.exit(1);
        }
        
        console.log("EN mode correctly rejects bilingual report");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_build_language_directive_v2_en_mode(self):
        """Test: buildLanguageDirectiveV2 with includeTranslation=false produces English-only directive"""
        test_code = '''
        import { buildLanguageDirectiveV2 } from "./src/lib/prompt-composer.js";
        
        const directive = buildLanguageDirectiveV2({
            inputDetected: "EN",
            outputEffective: "EN",
            mode: "final_report",
            includeTranslation: false
        });
        
        if (!directive.includes("English only")) {
            console.error("Expected 'English only' in directive, got:", directive);
            process.exit(1);
        }
        
        if (directive.includes("--- TRANSLATION ---")) {
            console.error("EN mode should not mention translation separator");
            process.exit(1);
        }
        
        console.log("buildLanguageDirectiveV2 correctly produces English-only directive");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_build_language_directive_v2_ru_mode(self):
        """Test: buildLanguageDirectiveV2 with includeTranslation=true produces translation directive"""
        test_code = '''
        import { buildLanguageDirectiveV2 } from "./src/lib/prompt-composer.js";
        
        const directive = buildLanguageDirectiveV2({
            inputDetected: "RU",
            outputEffective: "EN",
            mode: "final_report",
            includeTranslation: true,
            translationLanguage: "RU"
        });
        
        if (!directive.includes("--- TRANSLATION ---")) {
            console.error("Expected translation separator mention in directive, got:", directive);
            process.exit(1);
        }
        
        if (!directive.includes("translate the full output into Russian")) {
            console.error("Expected Russian translation instruction, got:", directive);
            process.exit(1);
        }
        
        console.log("buildLanguageDirectiveV2 correctly produces translation directive");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_prompt_files_no_hardcoded_translation(self):
        """Test: SYSTEM_PROMPT_BASE.txt does not contain hardcoded translation rules"""
        test_code = '''
        import { readFileSync } from "fs";
        import { join } from "path";
        
        const content = readFileSync(join(process.cwd(), "prompts/system/SYSTEM_PROMPT_BASE.txt"), "utf-8");
        
        if (content.includes("provide a full literal translation into the dialogue language")) {
            console.error("SYSTEM_PROMPT_BASE.txt should NOT contain hardcoded translation instruction");
            process.exit(1);
        }
        
        if (!content.includes("LANGUAGE DIRECTIVE")) {
            console.error("SYSTEM_PROMPT_BASE.txt should reference LANGUAGE DIRECTIVE");
            process.exit(1);
        }
        
        console.log("SYSTEM_PROMPT_BASE.txt correctly delegates to LANGUAGE DIRECTIVE");
        '''
        
        result = run_node_test(test_code)
        if not result["success"]:
            print(f"   Error: {result['stderr']}")
        return result["success"]

    def test_vitest_tests_pass(self):
        """Test: All existing vitest tests still pass"""
        print("Running complete vitest test suite...")
        result = subprocess.run(
            ["npx", "vitest", "run"], 
            cwd="/app",
            capture_output=True, 
            text=True,
            timeout=180
        )
        
        if result.returncode == 0:
            # Extract test count from output
            lines = result.stdout.split('\n')
            for line in lines:
                if "Tests" in line and "passed" in line:
                    print(f"   {line.strip()}")
            print("All vitest tests pass")
            return True
        else:
            print(f"   Error: {result.stderr[:300]}")
            return False

    def run_all_tests(self):
        """Run all language policy tests"""
        print("🚀 Starting RV Service Desk Language Policy Backend Tests")
        print("=" * 60)
        
        # Test language policy resolution
        self.test("resolveLanguagePolicy: EN mode → includeTranslation=false", 
                 self.test_resolve_language_policy_en_mode)
        
        self.test("resolveLanguagePolicy: RU mode → includeTranslation=true", 
                 self.test_resolve_language_policy_ru_mode)
        
        self.test("resolveLanguagePolicy: AUTO+EN → includeTranslation=false", 
                 self.test_resolve_language_policy_auto_en)
        
        self.test("resolveLanguagePolicy: AUTO+RU → includeTranslation=true", 
                 self.test_resolve_language_policy_auto_ru)
        
        # Test final report validation
        self.test("validateFinalReportOutput: EN mode passes English-only", 
                 self.test_validate_final_report_en_mode_english_only)
        
        self.test("validateFinalReportOutput: EN mode rejects bilingual", 
                 self.test_validate_final_report_en_mode_with_translation)
        
        # Test prompt composition
        self.test("buildLanguageDirectiveV2: EN mode → English-only directive", 
                 self.test_build_language_directive_v2_en_mode)
        
        self.test("buildLanguageDirectiveV2: RU mode → translation directive", 
                 self.test_build_language_directive_v2_ru_mode)
        
        # Test prompt files
        self.test("SYSTEM_PROMPT_BASE.txt: no hardcoded translation rules", 
                 self.test_prompt_files_no_hardcoded_translation)
        
        # Test complete vitest suite
        self.test("All 353 existing vitest tests still pass", 
                 self.test_vitest_tests_pass)
        
        # Print results
        print("\n" + "=" * 60)
        print(f"📊 Backend Test Results: {self.tests_passed}/{self.tests_run} passed")
        
        if self.failures:
            print(f"❌ Failed tests:")
            for failure in self.failures:
                print(f"  - {failure}")
            return 1
        else:
            print("✅ All backend tests passed!")
            return 0

def main():
    # Change to app directory
    os.chdir("/app")
    
    tester = LanguagePolicyTester()
    exit_code = tester.run_all_tests()
    
    return exit_code

if __name__ == "__main__":
    sys.exit(main())

File: tests/language-policy.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";

/**
 * Tests for declarative LanguagePolicy architecture.
 *
 * Ensures:
 * - resolveLanguagePolicy produces correct policy for every mode
 * - EN mode never includes translation
 * - RU / ES / AUTO (non-EN) always include translation
 * - Validators enforce the policy at output layer
 * - Prompt composer uses policy, not ad-hoc logic
 */

// ── resolveLanguagePolicy ──────────────────────────────────────────

describe("resolveLanguagePolicy", () => {
  beforeEach(() => { vi.resetModules(); });

  it("EN mode → no translation", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("EN", "EN");

    expect(policy.mode).toBe("EN");
    expect(policy.primaryOutput).toBe("EN");
    expect(policy.includeTranslation).toBe(false);
    expect(policy.translationLanguage).toBeUndefined();
  });

  it("EN mode with RU detected input → still no translation", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("EN", "RU");

    expect(policy.includeTranslation).toBe(false);
  });

  it("RU mode → translation into Russian", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("RU", "RU");

    expect(policy.mode).toBe("RU");
    expect(policy.primaryOutput).toBe("EN");
    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("RU");
  });

  it("ES mode → translation into Spanish", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("ES", "ES");

    expect(policy.mode).toBe("ES");
    expect(policy.primaryOutput).toBe("EN");
    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("ES");
  });

  it("AUTO mode + EN detected → no translation", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("AUTO", "EN");

    expect(policy.mode).toBe("AUTO");
    expect(policy.includeTranslation).toBe(false);
    expect(policy.translationLanguage).toBeUndefined();
  });

  it("AUTO mode + RU detected → translation into Russian", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("AUTO", "RU");

    expect(policy.mode).toBe("AUTO");
    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("RU");
  });

  it("AUTO mode + ES detected → translation into Spanish", async () => {
    const { resolveLanguagePolicy } = await import("@/lib/lang");
    const policy = resolveLanguagePolicy("AUTO", "ES");

    expect(policy.mode).toBe("AUTO");
    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("ES");
  });
});

// ── Validator enforcement ──────────────────────────────────────────

describe("validateFinalReportOutput – policy enforcement", () => {
  beforeEach(() => { vi.resetModules(); });

  const ENGLISH_ONLY_REPORT = [
    "Complaint: Water pump not operating per spec when activated.",
    "Diagnostic Procedure: Diagnostic checks included voltage verification at pump terminals.",
    "Verified Condition: Pump receives power but produces no flow.",
    "Recommended Corrective Action: Replace water pump assembly.",
    "Required Parts: Water pump assembly.",
    "Estimated Labor: Remove and replace pump - 1.5 hr. Total labor: 1.5 hr.",
  ].join("\n");

  const BILINGUAL_REPORT = [
    ENGLISH_ONLY_REPORT,
    "--- TRANSLATION ---",
    "Жалоба: Водяной насос не работает при активации.",
    "Диагностическая процедура: Проверено напряжение на клеммах насоса.",
    "Подтверждённое состояние: Насос получает питание, но нет потока.",
    "Рекомендованное корректирующее действие: Заменить узел водяного насоса.",
    "Оценка трудоёмкости: Снятие и замена — 1.5 ч. Общее время: 1.5 ч.",
    "Требуемые детали: Узел водяного насоса.",
  ].join("\n");

  it("EN mode (includeTranslation=false) → English-only report is valid", async () => {
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");
    const result = validateFinalReportOutput(ENGLISH_ONLY_REPORT, false);

    expect(result.valid).toBe(true);
    expect(result.violations).toHaveLength(0);
  });

  it("EN mode → report with translation block is INVALID", async () => {
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");
    const result = validateFinalReportOutput(BILINGUAL_REPORT, false);

    expect(result.valid).toBe(false);
    expect(result.violations.some(v => v.includes("EN mode must not include"))).toBe(true);
  });

  it("RU mode (includeTranslation=true) → bilingual report is valid", async () => {
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");
    const result = validateFinalReportOutput(BILINGUAL_REPORT, true, "RU");

    expect(result.valid).toBe(true);
  });

  it("RU/ES mode → English-only report is INVALID (missing translation)", async () => {
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");
    const result = validateFinalReportOutput(ENGLISH_ONLY_REPORT, true);

    expect(result.valid).toBe(false);
    expect(result.violations.some(v => v.includes("Missing '--- TRANSLATION ---'"))).toBe(true);
  });

  it("default (no includeTranslation arg) → backward compat: requires translation", async () => {
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");
    const result = validateFinalReportOutput(ENGLISH_ONLY_REPORT);

    expect(result.valid).toBe(false);
    expect(result.violations.some(v => v.includes("Missing '--- TRANSLATION ---'"))).toBe(true);
  });
});

describe("validateOutput dispatcher – forwards includeTranslation", () => {
  beforeEach(() => { vi.resetModules(); });

  it("final_report + includeTranslation=false → allows English-only", async () => {
    const { validateOutput } = await import("@/lib/mode-validators");
    const englishOnly = `Complaint: Water pump not operating per spec.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;

    const result = validateOutput(englishOnly, "final_report", false);
    expect(result.valid).toBe(true);
  });

  it("final_report + includeTranslation=true → rejects English-only", async () => {
    const { validateOutput } = await import("@/lib/mode-validators");
    const englishOnly = `Complaint: Water pump not operating per spec.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;

    const result = validateOutput(englishOnly, "final_report", true);
    expect(result.valid).toBe(false);
  });

  it("diagnostic mode ignores includeTranslation", async () => {
    const { validateOutput } = await import("@/lib/mode-validators");

    const result = validateOutput("Is the pump working?", "diagnostic", false);
    expect(result.valid).toBe(true);
  });
});

// ── Prompt composer – policy-driven directives ─────────────────────

describe("buildLanguageDirectiveV2 – policy-driven", () => {
  beforeEach(() => { vi.resetModules(); });

  it("final_report + includeTranslation=false → no translation mention", async () => {
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");

    const directive = buildLanguageDirectiveV2({
      inputDetected: "EN",
      outputEffective: "EN",
      mode: "final_report",
      includeTranslation: false,
    });

    expect(directive).toContain("English only");
    expect(directive).not.toContain("--- TRANSLATION ---");
    expect(directive).not.toContain("translate the full output");
  });

  it("final_report + includeTranslation=true → translation to RU", async () => {
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");

    const directive = buildLanguageDirectiveV2({
      inputDetected: "RU",
      outputEffective: "EN",
      mode: "final_report",
      includeTranslation: true,
      translationLanguage: "RU",
    });

    expect(directive).toContain("--- TRANSLATION ---");
    expect(directive).toContain("translate the full output into Russian (RU)");
  });

  it("final_report + includeTranslation=true → translation to ES", async () => {
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");

    const directive = buildLanguageDirectiveV2({
      inputDetected: "ES",
      outputEffective: "EN",
      mode: "final_report",
      includeTranslation: true,
      translationLanguage: "ES",
    });

    expect(directive).toContain("translate the full output into Spanish (ES)");
  });

  it("backward compat: no policy args → falls back to no-translation for EN input", async () => {
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");

    const directive = buildLanguageDirectiveV2({
      inputDetected: "EN",
      outputEffective: "EN",
      mode: "final_report",
    });

    // Without includeTranslation, defaults to no-translation path
    expect(directive).toContain("English only");
    expect(directive).not.toContain("--- TRANSLATION ---");
  });
});

// ── Output-layer enforcement ───────────────────────────────────────

describe("Output-layer enforcement (simulated)", () => {
  it("strips translation block when includeTranslation is false", () => {
    const SEPARATOR = "--- TRANSLATION ---";
    const text = `English report content.\n\n${SEPARATOR}\n\nRussian translation.`;
    const includeTranslation = false;

    let enforced = text;
    if (!includeTranslation && enforced.includes(SEPARATOR)) {
      enforced = enforced.split(SEPARATOR)[0].trim();
    }

    expect(enforced).toBe("English report content.");
    expect(enforced).not.toContain(SEPARATOR);
    expect(enforced).not.toContain("Russian");
  });

  it("preserves translation block when includeTranslation is true", () => {
    const SEPARATOR = "--- TRANSLATION ---";
    const text = `English report content.\n\n${SEPARATOR}\n\nRussian translation.`;
    const includeTranslation = true;

    let enforced = text;
    if (!includeTranslation && enforced.includes(SEPARATOR)) {
      enforced = enforced.split(SEPARATOR)[0].trim();
    }

    expect(enforced).toContain(SEPARATOR);
    expect(enforced).toContain("Russian translation.");
  });
});

// ── End-to-end policy scenarios ────────────────────────────────────

describe("End-to-end: mode → policy → directive → validation", () => {
  beforeEach(() => { vi.resetModules(); });

  it("EN mode → English-only pipeline", async () => {
    const { resolveLanguagePolicy, detectInputLanguageV2 } = await import("@/lib/lang");
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");

    const input = detectInputLanguageV2("Water pump not working");
    const policy = resolveLanguagePolicy("EN", input.detected);

    // Directive should NOT mention translation
    const directive = buildLanguageDirectiveV2({
      inputDetected: input.detected,
      outputEffective: policy.primaryOutput,
      mode: "final_report",
      includeTranslation: policy.includeTranslation,
      translationLanguage: policy.translationLanguage,
    });
    expect(directive).not.toContain("--- TRANSLATION ---");

    // English-only report should pass validation
    const report = `Complaint: Water pump not operating per spec.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;
    const validation = validateFinalReportOutput(report, policy.includeTranslation);
    expect(validation.valid).toBe(true);
  });

  it("RU mode → bilingual pipeline", async () => {
    const { resolveLanguagePolicy, detectInputLanguageV2 } = await import("@/lib/lang");
    const { buildLanguageDirectiveV2 } = await import("@/lib/prompt-composer");
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");

    const input = detectInputLanguageV2("Насос не работает");
    const policy = resolveLanguagePolicy("RU", input.detected);

    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("RU");

    // Directive should mention translation
    const directive = buildLanguageDirectiveV2({
      inputDetected: input.detected,
      outputEffective: policy.primaryOutput,
      mode: "final_report",
      includeTranslation: policy.includeTranslation,
      translationLanguage: policy.translationLanguage,
    });
    expect(directive).toContain("translate the full output into Russian");

    // English-only report should FAIL validation
    const englishOnly = `Complaint: Water pump not operating per spec.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;
    const v1 = validateFinalReportOutput(englishOnly, policy.includeTranslation);
    expect(v1.valid).toBe(false);

    // Bilingual report should pass
    const bilingual = englishOnly + "\n\n--- TRANSLATION ---\n\nНасос не работает. Работа: 1.0 час.";
    const v2 = validateFinalReportOutput(bilingual, policy.includeTranslation, policy.translationLanguage);
    expect(v2.valid).toBe(true);
  });

  it("AUTO + EN input → English-only pipeline", async () => {
    const { resolveLanguagePolicy, detectInputLanguageV2 } = await import("@/lib/lang");
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");

    const input = detectInputLanguageV2("The water pump is not producing flow");
    const policy = resolveLanguagePolicy("AUTO", input.detected);

    expect(input.detected).toBe("EN");
    expect(policy.includeTranslation).toBe(false);

    const report = `Complaint: Water pump not operating per spec.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;
    const validation = validateFinalReportOutput(report, policy.includeTranslation);
    expect(validation.valid).toBe(true);
  });

  it("AUTO + RU input → bilingual pipeline", async () => {
    const { resolveLanguagePolicy, detectInputLanguageV2 } = await import("@/lib/lang");
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");

    const input = detectInputLanguageV2("Клиент заявляет что насос не работает");
    const policy = resolveLanguagePolicy("AUTO", input.detected);

    expect(input.detected).toBe("RU");
    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("RU");

    const englishOnly = `Complaint: Water pump not operating.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.`;
    const v = validateFinalReportOutput(englishOnly, policy.includeTranslation);
    expect(v.valid).toBe(false);
  });

  it("ES mode → bilingual pipeline", async () => {
    const { resolveLanguagePolicy, detectInputLanguageV2 } = await import("@/lib/lang");
    const { validateFinalReportOutput } = await import("@/lib/mode-validators");

    const input = detectInputLanguageV2("¿Cómo está el diagnóstico?");
    const policy = resolveLanguagePolicy("ES", input.detected);

    expect(policy.includeTranslation).toBe(true);
    expect(policy.translationLanguage).toBe("ES");

    const bilingual = `Complaint: Water pump not operating.
Diagnostic Procedure: Verified voltage at pump terminals.
Verified Condition: Unit not responding under load.
Recommended Corrective Action: Replace pump.
Estimated Labor: Total labor: 1.0 hr.
Required Parts: Water pump assembly.

--- TRANSLATION ---

Queja: La bomba de agua no funciona.
Procedimiento de diagnóstico: Se verificó el voltaje en los terminales de la bomba.
Condición verificada: La unidad no responde bajo carga.
Acción correctiva recomendada: Reemplazar la bomba.
Mano de obra estimada: Total mano de obra: 1.0 hr.
Piezas requeridas: Bomba de agua.`;
    const v = validateFinalReportOutput(bilingual, policy.includeTranslation, policy.translationLanguage);
    expect(v.valid).toBe(true);
  });
});

// ── Prompt files: no hardcoded language enforcement ────────────────

describe("Prompt files: no hardcoded translation enforcement", () => {
  it("SYSTEM_PROMPT_BASE.txt should NOT hardcode 'provide a full literal translation'", async () => {
    const { readFileSync } = await import("fs");
    const { join } = await import("path");
    const content = readFileSync(join(process.cwd(), "prompts/system/SYSTEM_PROMPT_BASE.txt"), "utf-8");

    // Should NOT contain the old hardcoded instruction
    expect(content).not.toContain("provide a full literal translation into the dialogue language");
    // Should delegate to the LANGUAGE DIRECTIVE
    expect(content).toContain("LANGUAGE DIRECTIVE");
  });

  it("MODE_PROMPT_FINAL_REPORT.txt should conditionally reference translation", async () => {
    const { readFileSync } = await import("fs");
    const { join } = await import("path");
    const content = readFileSync(join(process.cwd(), "prompts/modes/MODE_PROMPT_FINAL_REPORT.txt"), "utf-8");

    // Should reference LANGUAGE DIRECTIVE for translation decision
    expect(content).toContain("LANGUAGE DIRECTIVE");
    // Should still contain the separator in the examples
    expect(content).toContain("--- TRANSLATION ---");
  });
});
